src/librustc_const_eval/us.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use syntax::ast;
+use super::err::*;
+
+/// Depending on the target only one variant is ever used in a compilation.
+/// Anything else is an error. This invariant is checked at several locations
+#[derive(Copy, Clone, Debug, RustcEncodable, RustcDecodable, Hash, Eq, PartialEq)]
+pub enum ConstUsize {
+    Us32(u32),
+    Us64(u64),
+}
+pub use self::ConstUsize::*;
+
+impl ConstUsize {
+    pub fn as_u64(self, target_uint_ty: ast::UintTy) -> u64 {
```

Comment:
I'm very unhappy with this function. The code is littered with situations where I want a `u64`, but only if it's actually fitting into a `usize` (whenever I pass this to llvm), and with situations where I want a `usize` (e.g. to const index an array), so it must fit into a `usize`. I had an `as_usize` method before, but that one panicked if the value didn't fit into the target's `usize`, which might be unexpected of such a simple function. I'm not sure how to proceed. I can add a FIXME and we fix it if it ever becomes a problem.